File: java/src/main/java/S056.java
import java.math.BigInteger;

public class S056 {

    private static final int LIMIT = 100;

    public static int solve() {
        int maxDigitSum = 0;

        // do not consider single digit values for a
        for (int a = 11; a < LIMIT; a++) {
            // probabilistically speaking, the max should be  200 * (5) ~ 1000
            // This is impossible if we have less than 100 digits, so consider b >= 50
            for (int b = 50; b < LIMIT; b++) {
                maxDigitSum = Math.max(maxDigitSum, digitSum(BigInteger.valueOf(a).pow(b)));
            }
        }
        return maxDigitSum;
    }

    private static int digitSum(BigInteger n) {
        String str = n.toString();
        int sum = 0;
        for (int i = 0; i < str.length(); i++) {
            sum += str.charAt(i) - '0';
        }
        return sum;
    }
}

File: cpp/S056.cpp
#include <iostream>
#include "biginteger.cpp"
#include "strings.cpp"

using namespace std;

// given
static const int LIMIT = 100;

int main() {
    int max_digit_sum = -1;
    for (int a = 1; a < LIMIT; a++) {
        // use base 10 for faster digit sums
        BigInteger big(a, 10);
        max_digit_sum = max(max_digit_sum, (int) big.sum_base());
        for (int b = 1; b < LIMIT; b++) {
            big.times(a);
            max_digit_sum = max(max_digit_sum, (int) big.sum_base());
        }
    }
    cout << max_digit_sum << endl;
    return 0;
}
